src/client/QXmppOmemoManager.cpp

```diff
@@ -0,0 +1,5001 @@
+// SPDX-FileCopyrightText: 2022 Melvin Keskin <melvo@olomono.de>
+// SPDX-FileCopyrightText: 2022 Linus Jahn <lnj@kaidan.im>
+//
+// SPDX-License-Identifier: LGPL-2.1-or-later
+
+#include "QXmppOmemoManager.h"
+
+#include "QXmppClient.h"
+#include "QXmppConstants_p.h"
+#include "QXmppOmemoDeviceElement_p.h"
+#include "QXmppOmemoElement_p.h"
+#include "QXmppOmemoEnvelope_p.h"
+#include "QXmppOmemoItems_p.h"
+#include "QXmppOmemoStorage.h"
+#include "QXmppPubSubEvent.h"
+#include "QXmppPubSubManager.h"
+#include "QXmppSceEnvelope_p.h"
+#include "QXmppStreamFeatures.h"
+#include "QXmppTrustManager.h"
+#include "QXmppUtils.h"
+#include <QXmppOmemoIq_p.h>
+
+#include <QCryptographicHash>
+#include <QDomElement>
+#include <QFutureSynchronizer>
+#include <QTimer>
+#if QT_VERSION >= QT_VERSION_CHECK(5, 10, 0)
+#include <QRandomGenerator>
+#endif
+#include <cstdint>
+#include <fingerprint.h>
+#include <key_helper.h>
+#include <protocol.h>
+#include <session_builder.h>
+#include <session_cipher.h>
+#include <signal_protocol.h>
+#include <signal_protocol_types.h>
+#include <stddef.h>
+
+#include <QStringBuilder>
+
+using namespace std::chrono_literals;
+
+using namespace QXmpp::Private;
+using Error = QXmppStanza::Error;
+
+// default label used for the own device
+const char *DEVICE_LABEL = "QXmpp";
+
+// count of unresponded stanzas sent to a device until QXmpp stops encrypting
+// for it
+constexpr int UNRESPONDED_STANZAS_UNTIL_ENCRYPTION_IS_STOPPED = 106;
+
+// count of unresponded stanzas received from a device until a heartbeat message
+// is sent to it
+constexpr int UNRESPONDED_STANZAS_UNTIL_HEARTBEAT_MESSAGE_IS_SENT = 53;
+
+// size of empty OMEMO message's decryption data
+constexpr int EMPTY_MESSAGE_DECRYPTION_DATA_SIZE = 32;
+
+// workaround for PubSub nodes that are not configurable to store 'max' as the
+// value for 'pubsub#max_items'
+constexpr uint64_t PUBSUB_NODE_MAX_ITEMS_1 = 1000;
+constexpr uint64_t PUBSUB_NODE_MAX_ITEMS_2 = 100;
+constexpr uint64_t PUBSUB_NODE_MAX_ITEMS_3 = 10;
```

Comment:
This is fine, just want to mention that e.g. the following would also be possible: `constexpr std::array PUBSUB_NODE_MAX_ITEMS_COUNTS = {1000, 100, 10};`

Reply:
Yes, the problem are the `await`s. Otherwise, it could be done much simpler via loops.